core-bundle/src/Routing/Page/PageRegistry.php

```diff
@@ -0,0 +1,198 @@
+
+class PageRegistry
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var array<RouteConfig>
+     */
+    private $routeConfigs = [];
+
+    /**
+     * @var array<DynamicRouteInterface>
+     */
+    private $routeEnhancers = [];
+
+    /**
+     * @var array<ContentCompositionInterface|bool>
+     */
+    private $contentComposition = [];
+
+    /**
+     * @var array<string>|null
+     */
+    private $urlPrefixes;
+
+    /**
+     * @var array<string>|null
+     */
+    private $urlSuffixes;
+
+    public function __construct(Connection $connection)
+    {
+        $this->connection = $connection;
+    }
+
+    public function getRouteConfig(string $type): RouteConfig
+    {
+        return $this->routeConfigs[$type] ?? new RouteConfig();
+    }
+
+    public function getPathRegex(): array
+    {
+        $prefixes = [];
+
+        foreach ($this->routeConfigs as $type => $config) {
+            $regex = $config->getPathRegex();
+
+            if (null !== $regex) {
+                $prefixes[$type] = $regex;
+            }
+        }
+
+        return $prefixes;
+    }
+
+    public function enhancePageRoute(PageRoute $route): Route
+    {
+        $type = $route->getPageModel()->type;
+
+        if (!isset($this->routeEnhancers[$type])) {
+            return $route;
+        }
+
+        /** @var DynamicRouteInterface $enhancer */
+        $enhancer = $this->routeEnhancers[$type];
+
+        return $enhancer->enhancePageRoute($route);
+    }
+
+    public function supportsContentComposition(PageModel $pageModel): bool
+    {
+        if (!isset($this->contentComposition[$pageModel->type])) {
+            return true;
+        }
+
+        $service = $this->contentComposition[$pageModel->type];
+
+        if ($service instanceof ContentCompositionInterface) {
+            return $service->supportsContentComposition($pageModel);
+        }
+
+        return (bool) $service;
+    }
+
+    /**
+     * @return array<string>
+     */
+    public function getUrlPrefixes(): array
+    {
+        $this->initializePrefixAndSuffix();
+
+        return $this->urlPrefixes;
+    }
+
+    /**
+     * @return array<string>
+     */
+    public function getUrlSuffixes(): array
+    {
+        $this->initializePrefixAndSuffix();
+
+        return $this->urlSuffixes;
+    }
+
+    /**
+     * @param ContentCompositionInterface|bool $contentComposition
+     */
+    public function add(string $type, RouteConfig $config, DynamicRouteInterface $routeEnhancer = null, $contentComposition = true): self
+    {
+        // Override existing pages with the same identifier
+        $this->routeConfigs[$type] = $config;
+
+        if (null !== $routeEnhancer) {
+            $this->routeEnhancers[$type] = $routeEnhancer;
+        }
+
+        if (null !== $contentComposition) {
+            $this->contentComposition[$type] = $contentComposition;
+        }
+
+        $this->urlPrefixes = $this->urlSuffixes = null;
```

Comment:
Just making sure: You intentionally set the properties to `null` here instead of `[]`, right?

Reply:
yes, because they (might) need to be re-initialized if the configuration changes.